sdk/ai/ai-projects/review/ai-projects.api.md

```diff
@@ -0,0 +1,1988 @@
+    event: AgentStreamEventType | string;
+}
+
+// @public
+export interface AgentEventMessageStream extends AsyncDisposable, AsyncIterable<AgentEventMessage> {
+}
+
+// @public
+export interface AgentOutput {
+    created_at: number;
+    description: string | null;
+    id: string;
+    instructions: string | null;
+    metadata: Record<string, string> | null;
+    model: string;
+    name: string | null;
+    object: "assistant";
+    response_format?: AgentsApiResponseFormatOptionOutput | null;
+    temperature: number | null;
+    tool_resources: ToolResourcesOutput | null;
+    tools: Array<ToolDefinitionOutput>;
+    top_p: number | null;
+}
+
+// @public
+export type AgentRunResponse = PromiseLike<ThreadRunOutput> & {
+    stream: () => Promise<AgentEventMessageStream>;
+};
+
+// @public
+export interface AgentsApiResponseFormat {
+    type?: ApiResponseFormat;
+}
+
+// @public
+export type AgentsApiResponseFormatMode = string;
+
+// @public
+export type AgentsApiResponseFormatModeOutput = string;
+
+// @public
+export type AgentsApiResponseFormatOption = string | AgentsApiResponseFormatMode | AgentsApiResponseFormat;
+
+// @public
+export type AgentsApiResponseFormatOptionOutput = string | AgentsApiResponseFormatModeOutput | AgentsApiResponseFormatOutput;
+
+// @public
+export interface AgentsApiResponseFormatOutput {
+    type?: ApiResponseFormatOutput;
+}
+
+// @public
+export type AgentsApiToolChoiceOption = string | AgentsApiToolChoiceOptionMode | AgentsNamedToolChoice;
+
+// @public
+export type AgentsApiToolChoiceOptionMode = string;
+
+// @public
+export type AgentsApiToolChoiceOptionModeOutput = string;
+
+// @public
+export type AgentsApiToolChoiceOptionOutput = string | AgentsApiToolChoiceOptionModeOutput | AgentsNamedToolChoiceOutput;
+
+// @public
+export interface AgentsNamedToolChoice {
+    function?: FunctionName;
+    type: AgentsNamedToolChoiceType;
+}
+
+// @public
+export interface AgentsNamedToolChoiceOutput {
+    function?: FunctionNameOutput;
+    type: AgentsNamedToolChoiceTypeOutput;
+}
+
+// @public
+export type AgentsNamedToolChoiceType = string;
+
+// @public
+export type AgentsNamedToolChoiceTypeOutput = string;
+
+// @public (undocumented)
+export interface AgentsOperations {
+    cancelRun: (threadId: string, runId: string, requestParams?: OptionalRequestParameters) => Promise<ThreadRunOutput>;
+    cancelVectorStoreFileBatch: (vectorStoreId: string, batchId: string, requestParams?: OptionalRequestParameters) => Promise<VectorStoreFileBatchOutput>;
+    createAgent: (model: string, options?: Omit<CreateAgentOptions, "model">, requestParams?: OptionalRequestParameters) => Promise<AgentOutput>;
+    createMessage: (threadId: string, options: ThreadMessageOptions, requestParams?: OptionalRequestParameters) => Promise<ThreadMessageOutput>;
+    createRun: (threadId: string, assistantId: string, options?: Omit<CreateRunOptions, "assistant_id">, requestParams?: OptionalRequestParameters) => AgentRunResponse;
```

Comment:
Can we combine `options` and `requestParams` into a single options bag? That is the common pattern

Reply:
This was to difference between what is needed for the api vs what is needed for the request which is always optional.

Reply:
This
`createAgent: (
    model: string,
    options?: Omit<CreateAgentOptions, "model">,
    requestParams?: OptionalRequestParameters
  ) => Promise<AgentOutput>;
`
To
`createAgent: (
    model: string,
    creatAgentOptions?: Omit<CreateAgentOptions, "model">,
    options?: OptionalRequestParameters
  ) => Promise<AgentOutput>;
`
or
`
type CreateAgentOptionalPrameters = Omit<CreateAgentOptions, "model"> & OptionalRequestParameters
createAgent: (
    model: string,
    options?: CreateAgentOptionalPrameters 
  ) => Promise<AgentOutput>;`

This works great for API that has everything in the body.
For API that need to take input and pass them as body or URL, this needs to be futher process, so I was leaning towards separating the service parameters with the request parameters.

Reply:
I gotcha. It gets awkward if a user wants to pass one but not the other:

```ts
createRun("threadId", "assistantId", undefined, { requestParams })
```

I get that it may require further processing at the convenience layer, but from a customer perspective this is the kind of API shape they will expect from Azure SDK client libraries. Mostly I see types that extend `OperationOptions` and add any service optional params, then split them up in the convenience layer. You can find such examples in [our repository](https://github.com/search?q=repo%3AAzure%2Fazure-sdk-for-js+%22extends+OperationOptions%22&type=code). Since this is a convenience client I would expect to see the same API design applied. Does that make sense?